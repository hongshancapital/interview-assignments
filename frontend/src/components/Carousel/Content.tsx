import React, { ReactElement } from "react"
import styled from 'styled-components'

const Container = styled.div<{
  count: number
  currentIndex: number
}>`
  position: absolute;
  width: ${props => (props.count * 100) + '%'};
  height: 100%;
  top: 0;
  left: 0;
  transform: ${({count, currentIndex}) => {
    const offsetPercent = count > 0 ? (-100.0 * currentIndex / count) : 0
    return 'translateX(' + offsetPercent + '%)'
  }};
  transition: 0.5s;
`

const Item = styled.div<{count: number}>`
  display: inline-block;
  position: relative;
  width: ${props => (100.0 / props.count) + '%'};
  height: 100%;
  box-sizing: border-box;
  overflow: hidden
`

export const Content = ({ currentIndex, children }: {
  currentIndex: number
  children: ReactElement[]
}) => {
  const count = children.length
  return (
    <Container count={count} currentIndex={currentIndex}>
      {
        children.map((child, index) =>
          <Item count={count} key={index}>
            {child.props.children}
          </Item>
        )
      }
    </Container>
  )
}
