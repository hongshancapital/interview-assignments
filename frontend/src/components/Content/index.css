.phoneContent {
  padding-top: 150px;
  background: url('../../assets/iphone.png') no-repeat center #101010;
  width: 100%;
  height: 100vh;
  background-size: cover;
  background-position: center -100px;
  color: #FFF;
}

.phoneContent .secondrytitle {
 width: 440px;
}

.tabletContent {
  padding-top: 150px;
  background: url('../../assets/tablet.png') no-repeat center #fafafa;
  width: 100%;
  height: 100%;
  background-size: contain;
  background-position: center 220px;
}

.arpodsContent {
  padding-top: 150px;
  background: url('../../assets/airpods.png') no-repeat center #f2f2f4;
  width: 100%;
  height: 100vh;
  background-size: contain;
  background-position: center 350px;
}


.bigtitle {
  font-size: 48px;
  font-weight: bold;
  width: 1000px;
  margin: 0 auto;
  word-break: break-word;
  white-space: pre-wrap;
}


.secondrytitle {
  font-size: 32px;
  margin: 0 auto;
  word-break: break-word;
  white-space: pre-wrap;
}