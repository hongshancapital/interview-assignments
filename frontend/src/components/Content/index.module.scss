.content {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding-top: 14%;
  padding-bottom: 8%;
  box-sizing:border-box;
}

.context h1 {
  font-size: 60px;
  margin: 0;
  padding: 0 10%;
}

.context p {
  font-size: 32px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-top: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.img {
  flex: 1;
  height: 0;
  object-fit:contain;
}