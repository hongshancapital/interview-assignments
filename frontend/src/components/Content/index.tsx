import React from "react";
import classnames from 'classnames'
import styles from './index.module.scss'

export interface IProps {
  backgroundColor: string
  color: string
  img: string
  title: string | string[]
  subtitle: string | string[]
}

export function Content(props: IProps) {
  const {
    backgroundColor,
    color,
    img,
    title,
    subtitle,
  } = props
  const _subtitle = Array.isArray(subtitle) ? subtitle : subtitle ? [] : [subtitle]
  const _title = Array.isArray(title) ? title : [title]
  return (
    <div style={{ backgroundColor }} className={classnames({ [styles.content]: true, [styles.iphone]: true })}>
      <div className={styles.context} style={{ color }}>
        {
          _title.length ? _title.map((t) => <h1 key={t}>{t}</h1>) : null
        }
        {_subtitle.length ? (
          <p>
            {_subtitle.map((sub) => <span key={sub}>{sub}</span>)}
          </p>
        ) : null}
      </div>

      <img className={styles.img} alt="img" src={img} />
    </div>
  );
}