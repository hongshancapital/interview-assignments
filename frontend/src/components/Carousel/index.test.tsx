import React from "react";
import { render, screen, fireEvent } from '@testing-library/react'
import { Carousel } from ".";
import { Content } from "../Content";

const items = [
  {
    color: '#fff',
    backgroundColor: '#111',
    img: require('../../assets/iphone.png'),
    title: 'xPhone',
    subtitle: ['Lots to love.Less to spend.', 'Starting as $399.']
  },
  {
    color: '#000',
    backgroundColor: '#fafafa',
    img: require('../../assets/tablet.png'),
    title: 'Tablet',
    subtitle: ['Just the right amount of everything.']
  },
  {
    color: '#000',
    backgroundColor: '#f1f1f3',
    img: require('../../assets/airpods.png'),
    title: ['Buy a Tablet Or xPhone for college.', 'Get airPods.'],
    subtitle: []
  }
]

describe('carousel', () => {
  test('快照', () => {
    const { container } = render(
      <Carousel auto indicator control width={800} height={400}>
        {items.map((item, index) => <Content key={index} {...item} />)}
      </Carousel>
    )
    expect(container).toMatchSnapshot()
  })

  test('点击左侧按钮', async () => {
    render(
      <Carousel auto indicator control width={800} height={400}>
        {items.map((item, index) => <Content key={index} {...item} />)}
      </Carousel>
    )
    const ele = await screen.findByTestId('elements')
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(-800px, 0, 0)')
    const next = await screen.findByTestId('next')
    fireEvent.click(next)
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(-1600px, 0, 0)')
  })

  test('点击右侧按钮', async () => {
    render(
      <Carousel auto indicator control width={800} height={400}>
        {items.map((item, index) => <Content key={index} {...item} />)}
      </Carousel>
    )
    const ele = await screen.findByTestId('elements')
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(-800px, 0, 0)')
    const prev = await screen.findByTestId('prev')
    fireEvent.click(prev)
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(0px, 0, 0)')
  })

  test('下方导航点击', async () => {
    render(
      <Carousel auto indicator control width={800} height={400}>
        {items.map((item, index) => <Content key={index} {...item} />)}
      </Carousel>
    )
    const ele = await screen.findByTestId('elements')
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(-800px, 0, 0)')
    const indicator = await screen.findByTestId('indicator_1')
    fireEvent.click(indicator)
    expect(ele.style.getPropertyValue('transform')).toBe('translate3d(-1600px, 0, 0)')
  })
})