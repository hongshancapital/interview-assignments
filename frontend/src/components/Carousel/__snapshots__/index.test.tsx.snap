// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`carousel 快照 1`] = `
<div>
  <div
    class="carousel_box"
    style="width: 800px; height: 400px;"
  >
    <div
      class="carousel_elements"
      data-testid="elements"
      style="transform: translate3d(-800px, 0, 0);"
    >
      <div
        class="element"
        style="left: 0px;"
      >
        <div
          class="content iphone"
          style="background-color: rgb(241, 241, 243);"
        >
          <div
            class="context"
            style="color: rgb(0, 0, 0);"
          >
            <h1>
              Buy a Tablet Or xPhone for college.
            </h1>
            <h1>
              Get airPods.
            </h1>
          </div>
          <img
            alt="img"
            class="img"
            src="airpods.png"
          />
        </div>
      </div>
      <div
        class="element"
        style="left: 800px;"
      >
        <div
          class="content iphone"
          style="background-color: rgb(17, 17, 17);"
        >
          <div
            class="context"
            style="color: rgb(255, 255, 255);"
          >
            <h1>
              xPhone
            </h1>
            <p>
              <span>
                Lots to love.Less to spend.
              </span>
              <span>
                Starting as $399.
              </span>
            </p>
          </div>
          <img
            alt="img"
            class="img"
            src="iphone.png"
          />
        </div>
      </div>
      <div
        class="element"
        style="left: 1600px;"
      >
        <div
          class="content iphone"
          style="background-color: rgb(250, 250, 250);"
        >
          <div
            class="context"
            style="color: rgb(0, 0, 0);"
          >
            <h1>
              Tablet
            </h1>
            <p>
              <span>
                Just the right amount of everything.
              </span>
            </p>
          </div>
          <img
            alt="img"
            class="img"
            src="tablet.png"
          />
        </div>
      </div>
      <div
        class="element"
        style="left: 2400px;"
      >
        <div
          class="content iphone"
          style="background-color: rgb(241, 241, 243);"
        >
          <div
            class="context"
            style="color: rgb(0, 0, 0);"
          >
            <h1>
              Buy a Tablet Or xPhone for college.
            </h1>
            <h1>
              Get airPods.
            </h1>
          </div>
          <img
            alt="img"
            class="img"
            src="airpods.png"
          />
        </div>
      </div>
      <div
        class="element"
        style="left: 3200px;"
      >
        <div
          class="content iphone"
          style="background-color: rgb(17, 17, 17);"
        >
          <div
            class="context"
            style="color: rgb(255, 255, 255);"
          >
            <h1>
              xPhone
            </h1>
            <p>
              <span>
                Lots to love.Less to spend.
              </span>
              <span>
                Starting as $399.
              </span>
            </p>
          </div>
          <img
            alt="img"
            class="img"
            src="iphone.png"
          />
        </div>
      </div>
    </div>
    <div
      class="controller controller_prev"
      data-testid="prev"
      style="width: 32px; height: 32px;"
    />
    <div
      class="controller controller_next"
      data-testid="next"
      style="width: 32px; height: 32px;"
    />
    <div
      class="indicator"
    >
      <span
        class="indicator_item indicator_active"
        data-testid="indicator_0"
      >
        <b
          style="transition: all 3s linear;"
        />
      </span>
      <span
        class="indicator_item"
        data-testid="indicator_1"
      >
        <b
          style="transition: none;"
        />
      </span>
      <span
        class="indicator_item"
        data-testid="indicator_2"
      >
        <b
          style="transition: none;"
        />
      </span>
    </div>
  </div>
</div>
`;
